src/bit-systems/waypoint.ts

```diff
@@ -164,6 +171,16 @@ export function waypointSystem(
       }
     }
 
+    const hashUpdated = window.location.hash !== "" && previousWaypointHash !== window.location.hash;
+    if (hashUpdated && initialSpawnHappened) {
+      const waypointName = window.location.hash.replace("#", "");
+      if (eid) {
+        moveToWaypoint(world, eid, characterController, previousWaypointHash === null);
+        window.history.replaceState(null, "", window.location.href.split("#")[0]); // Reset so you can re-activate the same waypoint
+      }
+      previousWaypointHash = window.location.hash;
+    }
```

Comment:
This part is very confusing to me... Should it really be processed againts `waypointQuery`? There is a chance that there are multiple entities that have waypoint components but only the first entity can be processed because `previousWaypointHash !== window.location.hash` will be `false` for the next ones.

If it is intentional, the code should be clearer.

Reply:
Perhaps the code will be clearer if `hashUpdated` check is done outside of the loop and named waypoint is handled in another query loop.

```typescript
function export function waypointSystem(...) {
  const hashUpdated = window.location.hash !== "" && previousWaypointHash !== window.location.hash;
  if (hashUpdated && initialSpawnHappened) {
    const waypointName = window.location.hash.replace("#", "");
    waypointQuery(world).forEach(eid => {
      const waypointObj = world.eid2obj.get(eid);
      if (waypointObj && waypointObj.name === waypointName) {
        moveToWaypoint(world, eid, characterController, previousWaypointHash === null);
        window.history.replaceState(null, "", window.location.href.split("#")[0]); // Reset so you can re-activate the same waypoint
        previousWaypointHash = window.location.hash;
        break;
      }
    });
  }

  waypointQuery(world).forEach(eid => {
    // process others
    ...
  });
}
```

Reply:
Yes, that's reasonable, refactored.